src/writers/sink.creator.cpp

```diff
@@ -239,41 +252,141 @@ zarr::SinkCreator::make_metadata_sinks_(
-        dir_paths.pop();
-        if (!fs::is_directory(dir) && !fs::create_directories(dir)) {
+        for (auto i = 3; i < tokens.size() - 1; ++i) {
+            base_dir += tokens.at(i) + "/";
+        }
+        if (tokens.size() > 3) {
+            base_dir += tokens.at(tokens.size() - 1);
+        }
+
+        // create the bucket if it doesn't already exist
+        if (!bucket_exists_(bucket_name)) {
             return false;
         }
+    } else {
+        base_dir = base_uri;
+        if (base_uri.starts_with("file://")) {
+            base_dir = base_uri.substr(7);
+        }
+
+        // remove trailing slashes
+        if (base_uri.ends_with("/") || base_uri.ends_with("\\")) {
+            base_dir = base_dir.substr(0, base_dir.size() - 1);
+        }
+
+        // create the base directories if they don't already exist
+        // we create them in serial because
+        // 1. there are only a few of them; and
+        // 2. they may be nested
+        while (!dir_paths.empty()) {
+            const auto dir = base_dir + "/" + dir_paths.front();
+            dir_paths.pop();
+            if (!fs::is_directory(dir) && !fs::create_directories(dir)) {
+                return false;
+            }
+        }
     }
 
     // make files
     size_t n_paths = file_paths.size();
+    const std::string prefix = base_dir.empty() ? "" : base_dir + "/";
     for (auto i = 0; i < n_paths; ++i) {
-        const auto path = base_dir + "/" + file_paths.front();
+        const auto path = prefix + file_paths.front();
         file_paths.pop();
         file_paths.push(path);
     }
 
-    if (!make_files_(file_paths, metadata_sinks)) {
+    return is_s3 ? make_s3_objects_(bucket_name, file_paths, metadata_sinks)
+                 : make_files_(file_paths, metadata_sinks);
+}
+
+bool
+zarr::SinkCreator::bucket_exists_(const std::string& bucket_name)
+{
+    CHECK(!bucket_name.empty());
+    EXPECT(connection_pool_, "S3 connection pool not provided.");
+
+    auto conn = connection_pool_->get_connection();
+    bool bucket_exists = conn->bucket_exists(bucket_name);
+
+    connection_pool_->return_connection(std::move(conn));
+
+    return bucket_exists;
+}
+
+bool
+zarr::SinkCreator::make_s3_objects_(const std::string& bucket_name,
+                                    std::queue<std::string>& object_keys,
+                                    std::vector<std::unique_ptr<Sink>>& sinks)
+{
+    if (object_keys.empty()) {
+        return true;
+    }
+    if (!connection_pool_) {
+        LOGE("S3 connection pool not provided.");
         return false;
     }
 
-    return true;
+    std::atomic<bool> all_successful = true;
+
+    const auto n_objects = object_keys.size();
+    sinks.resize(n_objects);
+    std::fill(sinks.begin(), sinks.end(), nullptr);
+    std::latch latch(n_objects);
+
+    for (auto i = 0; i < n_objects; ++i) {
+        const auto object_key = object_keys.front();
+        object_keys.pop();
```

Comment:
- I use the queue to construct paths, as we discussed in a previous PR. I'd have to rework that whole thing. I really don't think that a vector is the right tool for constructing paths like `/(1...k)/(1...m)/(1...n)/.../(1...z)`.
- I've been thinking that since we're not actually doing anything resource intensive, like opening a file, when we create an S3Sink, we can do this serially.

Reply:
I think I’m missing something here. I’ll tinker with the code to better understand this requirement.